ratecounter.go

```diff
@@ -2,27 +2,68 @@ package ratecounter
 
 import (
 	"strconv"
+	"sync/atomic"
 	"time"
 )
 
 // A RateCounter is a thread-safe counter which returns the number of times
 // 'Incr' has been called in the last interval
 type RateCounter struct {
-	counter  Counter
-	interval time.Duration
+	counter    Counter
+	interval   time.Duration
+	resolution int
+	partials   []Counter
+	current    int
+	running    int32
 }
 
 // NewRateCounter Constructs a new RateCounter, for the interval provided
 func NewRateCounter(intrvl time.Duration) *RateCounter {
-	return &RateCounter{
+	ratecounter := &RateCounter{
 		interval: intrvl,
+		running:  0,
 	}
+
+	return ratecounter.WithResolution(20)
+}
+
+func (r *RateCounter) WithResolution(resolution int) *RateCounter {
```

Comment:
Because this resets everything, let's do it all at once with:
`func NewRateCounter(intrvl time.Duration, resolution int) *RateCounter`